src/main/scala/cloud/drdrdr/utils/AWSCredentials.scala

```diff
@@ -6,165 +6,416 @@ package cloud.drdrdr.utils
 
 
 import java.io.File
+import java.text.SimpleDateFormat
+import java.util.{Date, SimpleTimeZone}
 
 import akka.actor.ActorSystem
+import akka.http.scaladsl.Http
 import akka.http.scaladsl.model._
-import akka.stream.ActorMaterializer
-import akka.stream.scaladsl.Sink
 import akka.pattern.after
-import cloud.drdrdr.SignRequestForAWS
-import scala.io.Source
+import akka.stream.{ActorMaterializer, scaladsl}
+import akka.stream.scaladsl.Sink
+import akka.util.ByteString
 import spray.json._
+import org.slf4j.LoggerFactory
+import com.typesafe.scalalogging.Logger
 import scala.concurrent.duration._
-import scala.concurrent.{Promise, Await, ExecutionContext, Future}
-import scala.util.{Success, Failure}
+import scala.concurrent.{ExecutionContext, Future, Await}
+import scala.io.Source
+import scala.util.control
+
+
+trait AWSCredentials {
+
+  case class AWSCredentialSource(creds: Future[AWSPermissions]) (implicit ec: ExecutionContext, s:ActorSystem, m: ActorMaterializer) {
+    private var credentials: Future[AWSPermissions] = creds
+
+    /**
+     * Updates the credentials if necessary, then returns the credentials
+     * @return AWSPermissions with most recent credentials
+     */
+    def getCredentials: Future[AWSPermissions] = {
+      credentials flatMap {
+        case perm:AWSPermissions =>
+          if (perm.expiration.isEmpty) {
+              credentials
+          } else {
+              checkExpire flatMap {
+                case bol =>
+                  if (bol) {
+                    credentials = updateCredentials()
+                  }
+                  credentials
+              }
+          }
+      }
+    }
+
+    //checks if current credentials have expired
+    private def checkExpire()(implicit ec: ExecutionContext, s: ActorSystem, m: ActorMaterializer): Future[Boolean] = {
+      credentials map {
+        case creds: AWSPermissions =>
+          if (creds.expiration.isEmpty)
+            false
+          else
+            creds.expiration <= getUTCTime()
+      }
+    }
+//
+    //updates the credentials on an EC2 instance if necessary
+    @throws(classOf[Exception])
+    private def updateCredentials(): Future[AWSPermissions] = {
+      getAmazonEC2Credentials() map {
+        case Some(cred) => cred
+        case None => throw new Exception("Unable to update credentials")
```

Comment:
Does that matter because the return type of getAmazonEC2Credentials is Future[Option], so if you are able to map then it has to be Some() or None